File: Graphics21/src/main/java/org/workcraft/graphics/LabelPositioning.java
package org.workcraft.graphics;

import static pcollections.TreePVector.empty;

import java.awt.geom.AffineTransform;
import java.awt.geom.Rectangle2D;

import org.workcraft.util.Pair;

import pcollections.PVector;

public enum LabelPositioning {
	TOP("Top", 0, -1),
	LEFT("Left", -1, 0),
	RIGHT("Right", 1, 0),
	BOTTOM("Bottom", 0, 1),
	CENTER("Center", 0, 0);

	public final String name;
	public final int dx, dy;

	private LabelPositioning(String name, int dx, int dy) {
		this.name = name;
		this.dx = dx;
		this.dy = dy;
	}

	public static PVector<Pair<String, LabelPositioning>> getChoice() {
		PVector<Pair<String, LabelPositioning>> positions = empty();
		
		for(LabelPositioning lp : LabelPositioning.values())
			positions = positions.plus(Pair.of(lp.name, lp));
		
		return positions;
	}
	
	public static AffineTransform positionRelative(Rectangle2D what, Rectangle2D relativeTo, LabelPositioning positioning) {
		double tx = -what.getCenterX() + relativeTo.getCenterX() + 0.5 * positioning.dx * (relativeTo.getWidth() + what.getWidth() + 0.2);
		double ty = -what.getCenterY() + relativeTo.getCenterY() + 0.5 * positioning.dy * (relativeTo.getHeight() + what.getHeight() + 0.2);
		
		return AffineTransform.getTranslateInstance(tx, ty);
	}
}

File: Graphics/src/main/scala/org/workcraft/graphics/LabelPositioning.scala
package org.workcraft.graphics
import java.awt.geom.Rectangle2D
import java.awt.geom.AffineTransform

sealed abstract class LabelPositioning (val name: String, val dx: Int, val dy: Int)

object LabelPositioning {
  case object Top extends LabelPositioning ("Top", 0, -1) 
  case object Left extends LabelPositioning ("Left", -1, 0)
  case object Right extends LabelPositioning ("Right", 1, 0)
  case object Bottom extends LabelPositioning ("Bottom", 0, 1)
  case object Center extends LabelPositioning ("Center", 0, 0)
  
  def values = List(Top, Left, Right, Bottom, Center)
  
  def getChoice = values.map (v => (v.name, v))
  
  def positionRelative (what: Rectangle2D, relativeTo: Rectangle2D, positioning: LabelPositioning) = {
		val tx = -what.getCenterX() + relativeTo.getCenterX() + 0.5 * positioning.dx * (relativeTo.getWidth() + what.getWidth() + 0.2)
		val ty = -what.getCenterY() + relativeTo.getCenterY() + 0.5 * positioning.dy * (relativeTo.getHeight() + what.getHeight() + 0.2)
		
		AffineTransform.getTranslateInstance(tx, ty)
  }
}